### PR TITLE
pkg: archive: only ignore ENOTSUP when xattr fails

### PR DESCRIPTION
**Pull request by cyphar**
There might be other (valid) reasons for setxattr(2) to fail, so only
ignore it when it's a not supported error (ENOSUPP). Otherwise, bail.

Fixes #21251

Signed-off-by: Aleksa Sarai asarai@suse.de

/cc @rhatdan @runcom [Can you please check if this fixes the SELinux problem without also breaking AUFS again -- I don't have an AUFS box to test this on at the moment]

### CODE REVIEW

**Comment by runcom** [comment]
@cyphar will test it in a while - looking good though since `lsetxattr` is returning `ENOTSUP` from the strace I did (and unless golang changes it it looks good)

File: pkg/archive/archive.go [hunk 1]
```diff
@@ -428,8 +428,14 @@ func createTarFile(path, extractDir string, hdr *tar.Header, reader io.Reader, L
 	var errors []string
 	for key, value := range hdr.Xattrs {
 		if err := system.Lsetxattr(path, key, []byte(value), 0); err != nil {
-			// We ignore errors here because not all graphdrivers support xattrs.
-			errors = append(errors, err.Error())
+			if err == syscall.ENOTSUP {
+				// We ignore errors here because not all graphdrivers support
+				// xattrs *cough* AUFS *cough*. However only ENOTSUP should be
+				// emitted in that case.
```
Comment (justincormack) [inline pkg/archive/archive.go]: The comment is not strictly true, new versions of aufs do support xattr.

**Comment by justincormack** [comment]
@cyphar you said fixes #21251 but thats a PR not an issue...

**Comment by runcom** [comment]
ust tested on a rhel72

LGTM (after @justincormack comment is fixed)

**Comment by cyphar** [comment]
@justincormack "fixes" as in it fixes an issue that @runcom and @rhatdan noted in that PR.

File: pkg/archive/archive.go (hunk 1 above)
Comment (cyphar) [inline pkg/archive/archive.go]: Fixed.

File: pkg/archive/archive.go [hunk 2]
```diff
@@ -428,8 +428,15 @@ func createTarFile(path, extractDir string, hdr *tar.Header, reader io.Reader, L
 	var errors []string
 	for key, value := range hdr.Xattrs {
 		if err := system.Lsetxattr(path, key, []byte(value), 0); err != nil {
-			// We ignore errors here because not all graphdrivers support xattrs.
```
Comment (rhatdan) [inline pkg/archive/archive.go]: Why not just change to

```
if err := system.Lsetxattr(path, key, []byte(value), 0); err != nil && err != syscall.ENOTSUP {
```

Comment (runcom) [inline pkg/archive/archive.go]: probably for logging below

Comment (cyphar) [inline pkg/archive/archive.go]: Yeah, it's so that we have logging of the errors we ignored.

**Comment by thaJeztah** [comment]
LGTM

**Comment by justincormack** [comment]
LGTM. All green (gccgo has passed, somehow the hook was missed).

**Pull request merged**